csrc/flash_attn_ck/mha_fwd.cpp

```diff
@@ -0,0 +1,350 @@
+
+    TORCH_CHECK(q.stride(-1) == 1, "Input tensor must have contiguous last dimension");
+    TORCH_CHECK(k.stride(-1) == 1, "Input tensor must have contiguous last dimension");
+    TORCH_CHECK(v.stride(-1) == 1, "Input tensor must have contiguous last dimension");
+
+    const auto sizes = q.sizes();
+
+    const int batch_size = sizes[0];
+    int seqlen_q = sizes[1];
+    int num_heads = sizes[2];
+    const int head_size_og = sizes[3];
+    const int seqlen_k = k.size(1);
+    const int num_heads_k = k.size(2);
+    TORCH_CHECK(batch_size > 0, "batch size must be postive");
+    TORCH_CHECK(head_size_og <= 256, "CK only supports head dimension at most 256");
+    TORCH_CHECK(num_heads % num_heads_k == 0, "Number of heads in key/value must divide number of heads in query");
+
+    if (window_size_left >= seqlen_k)
+    {
+        window_size_left = -1;
+    }
+    if (window_size_right >= seqlen_k)
+    {
+        window_size_right = -1;
+    }
+
+    // causal=true is the same as causal=false in this case
+    if (seqlen_q == 1 && !alibi_slopes_.has_value())
+    {
+        is_causal = false;
+    }
+
+    mask_info mask;
+
+    if (is_causal)
+    {
+        // Causal is the special case where window_size_right == 0 and window_size_left < 0.
+        window_size_right = 0;
+        std::string mask_identify = "b:" + std::to_string(window_size_left) + "," + "0";
+        mask = mask_info::decode(mask_identify, seqlen_q, seqlen_k); // casual
+    }
+    else if (window_size_left == -1 && window_size_right == -1)
+    {
+        mask = mask_info::decode("0", seqlen_q, seqlen_k); // no mask
+    }
+    else
+    {
+        // Local is the more general case where window_size_right >= 0 or window_size_left >= 0.
+        std::string mask_identify = "b:" + std::to_string(window_size_left) + "," + std::to_string(window_size_right);
+        mask = mask_info::decode(mask_identify, seqlen_q, seqlen_k); // local
+    }
+
+    // TODO
+    // Faster to transpose q from (b, 1, (nheads_kv ngroups), d) to (b, ngroups, nheads_kv, d) in this case
+    // H/t Daniel Haziza
+
+    CHECK_SHAPE(q, batch_size, seqlen_q, num_heads, head_size_og);
+    CHECK_SHAPE(k, batch_size, seqlen_k, num_heads_k, head_size_og);
+    CHECK_SHAPE(v, batch_size, seqlen_k, num_heads_k, head_size_og);
+
+    at::Tensor q_padded, k_padded, v_padded;
+    if (head_size_og % 8 != 0)
+    {
+        q_padded = torch::nn::functional::pad(q, torch::nn::functional::PadFuncOptions({0, 8 - head_size_og % 8}));
+        k_padded = torch::nn::functional::pad(k, torch::nn::functional::PadFuncOptions({0, 8 - head_size_og % 8}));
+        v_padded = torch::nn::functional::pad(v, torch::nn::functional::PadFuncOptions({0, 8 - head_size_og % 8}));
+    }
+    else
+    {
+        q_padded = q;
+        k_padded = k;
+        v_padded = v;
+    }
+
+    auto round_multiple = [](int x, int m)
+    { return (x + m - 1) / m * m; };
+    const int head_size_8x = round_multiple(head_size_og, 8);
```

Comment:
could we move these 3 lines right after the `if (out_.has_value())` statement? the same order as the original _flash_api.cpp_

```c++
at::Tensor out;
if (out_.has_value()) {
   // ...
} else {
   // ...
}

auto round_multiple = [](int x, int m)
    { return (x + m - 1) / m * m; };
const int head_size_8x = round_multiple(head_size_og, 8);
```

Reply:
85692866ab88f006c6d7ce3c5009129007a70108